packages/lesswrong/server/search/elastic/ElasticSearchClient.ts

```diff
@@ -0,0 +1,67 @@
+import { Client } from "@elastic/elasticsearch";
+import type { SearchHit, SearchResponse } from "@elastic/elasticsearch/lib/api/types";
+import { PublicInstanceSetting } from "../../../lib/instanceSettings";
+import ElasticSearchQuery, { QueryData } from "./ElasticSearchQuery";
+
+const elasticCloudIdSetting = new PublicInstanceSetting<string|null>(
+  "elasticsearch.cloudId",
+  null,
+  "optional",
+);
+const elasticUsernameSetting = new PublicInstanceSetting<string|null>(
+  "elasticsearch.username",
+  null,
+  "optional",
+);
+const elasticPasswordSetting = new PublicInstanceSetting<string|null>(
+  "elasticsearch.password",
+  null,
+  "optional",
+);
+
+export type ElasticDocument = Exclude<AlgoliaDocument, "_id">;
+export type ElasticSearchHit = SearchHit<ElasticDocument>;
+export type ElasticSearchResponse = SearchResponse<ElasticDocument>;
+
+class ElasticSearchClient {
+  private client: Client;
```

Comment:
I think this should be `private client?: Client;`, to cover the case where the settings aren't configured?

(I gotta say I don't love a pattern which has us writing code where the class methods themselves aren't sure whether the client is initialized, and have to include logic like `getClientOrThrow`.  Can we just front-load that assumption to the constructor, which will throw an error if any of the settings aren't appropriately configured, and require the caller to have done due diligence for attempting to create an instance of the client in the first place?  I think this probably looks like "pass in the credentials to the constructor".  Not sure what to do about the case where the credentials exist but are invalid.  I assume calling `new Client` will still succeed, but you'll just get an error when later attempting to hit the API?)